tests/elasticarray/main.c

```diff
@@ -0,0 +1,167 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "elasticarray.h"
+
+#define NUM_ELEM 3
+
+struct stuff {
+	int i; 
+};
+
+struct dyn {
+	int * arr;
+};
+
+ELASTICARRAY_DECL(INTLIST, intlist, int);
+ELASTICARRAY_DECL(STUFFLIST, stufflist, struct stuff);
+ELASTICARRAY_DECL(DYNLIST, dynlist, struct dyn *);
+
+/* Elasticarray with simple integers. */
+static int
+test_intlist()
+{
+	INTLIST list;
+	size_t i;
+	int x;
+
+	/* Allocate list. */
+	if ((list = intlist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		x = (int)i;
+		if (intlist_append(list, &x, 1))
+			goto err0;
+	}
+
+	/* Use the list. */
+	for (i = 0; i < intlist_getsize(list); i++) {
+		x = *intlist_get(list, i);
+		printf("int:\t%i\n", x);
+	}
+
+	/* Free memory. */
+	intlist_free(list);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Elasticarray with simple structures. */
+static int
+test_stufflist()
+{
+	STUFFLIST list;
+	struct stuff s;
+	size_t i;
+
+	/* Allocate list. */
+	if ((list = stufflist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		s.i = (int)i;
+
+		/* Append item. */
+		if (stufflist_append(list, &s, 1))
+			goto err0;
+	}
+
+	/* Use the list. */
+	for (i = 0; i < stufflist_getsize(list); i++) {
+		s = *stufflist_get(list, i);
+		printf("stuff:\t%i\n", s.i);
+	}
+
+	/* Free memory. */
+	stufflist_free(list);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Elasticarray with dynamically-allocated structures. */
+static int
+test_dynlist()
+{
+	DYNLIST list;
+	struct dyn * p;
+	size_t i;
+
+	/* Allocate list. */
+	if ((list = dynlist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		/* Dynamically allocate item to append. */
+		if ((p = malloc(sizeof(struct dyn))) == NULL)
+			goto err0;
+		if ((p->arr = malloc(sizeof(int))) == NULL)
+			goto err0;
+		p->arr[0] = (int)i;
+
+		/* Append item. */
+		if (dynlist_append(list, &p, 1))
+			goto err0;
+	}
+
+	/* Use the list. */
+	for (i = 0; i < dynlist_getsize(list); i++) {
+		p = *dynlist_get(list, i);
+		printf("dyn:\t%i\n", p->arr[0]);
+	}
+
+	/* Free memory. */
```

Comment:
This part is the real reason I'm adding this test.  Whenever I try to use elasticarray for structs, I end up in a mess of memory leaks and/or invalid `free()`.  (and yes, one of my branches currently has both problems)

Having a simple, reliable, reference "use case" of elasticarray will be helpful for me.